File: src/BlazorPwaKit/wwwroot/serviceWorkerManager.js
// BlazorPwaKit Service Worker Manager JavaScript Module
let dotNetReference = null;
let registrations = new Map();

export function initialize(dotNetObjectReference) {
    dotNetReference = dotNetObjectReference;
    
    // Listen for service worker events if supported
    if ('serviceWorker' in navigator) {
        navigator.serviceWorker.addEventListener('message', (event) => {
            if (dotNetReference) {
                if (event.data && event.data.type === 'UPDATE_AVAILABLE') {
                    dotNetReference.invokeMethodAsync('OnServiceWorkerEvent', 'updateavailable', null);
                } else {
                    dotNetReference.invokeMethodAsync('OnServiceWorkerEvent', 'message', event.data);
                }
            }
        });
    }
}

export function isServiceWorkerSupported() {
    return 'serviceWorker' in navigator;
}

export async function registerServiceWorker(scriptUrl, scope) {
    if (!isServiceWorkerSupported()) {
        throw new Error('Service workers are not supported in this browser');
    }

    try {
        const options = scope ? { scope } : undefined;
        const registration = await navigator.serviceWorker.register(scriptUrl, options);
        
        // Store registration for later reference
        const key = `${scriptUrl}:${scope || '/'}`;
        registrations.set(key, registration);
        
        // Set up event listeners
        setupRegistrationEventListeners(registration);
        
        return true;
    } catch (error) {
        console.error('Service worker registration failed:', error);
        if (dotNetReference) {
            dotNetReference.invokeMethodAsync('OnServiceWorkerEvent', 'error', error.message);
        }
        return false;
    }
}

export async function updateServiceWorker(scriptUrl, scope) {
    if (!isServiceWorkerSupported()) {
        return false;
    }

    try {
        const key = `${scriptUrl}:${scope || '/'}`;
        let registration = registrations.get(key);
        
        if (!registration) {
            // Try to get existing registration
            registration = await navigator.serviceWorker.getRegistration(scope);
            if (registration) {
                registrations.set(key, registration);
            }
        }
        
        if (registration) {
            const updatedRegistration = await registration.update();
            if (dotNetReference) {
                dotNetReference.invokeMethodAsync('OnServiceWorkerEvent', 'updated', null);
            }
            return true;
        }
        
        return false;
    } catch (error) {
        console.error('Service worker update failed:', error);
        if (dotNetReference) {
            dotNetReference.invokeMethodAsync('OnServiceWorkerEvent', 'error', error.message);
        }
        return false;
    }
}

export async function unregisterServiceWorker(scriptUrl, scope) {
    if (!isServiceWorkerSupported()) {
        return false;
    }

    try {
        const key = `${scriptUrl}:${scope || '/'}`;
        let registration = registrations.get(key);
        
        if (!registration) {
            // Try to get existing registration
            registration = await navigator.serviceWorker.getRegistration(scope);
        }
        
        if (registration) {
            const success = await registration.unregister();
            if (success) {
                registrations.delete(key);
            }
            return success;
        }
        
        return false;
    } catch (error) {
        console.error('Service worker unregistration failed:', error);
        if (dotNetReference) {
            dotNetReference.invokeMethodAsync('OnServiceWorkerEvent', 'error', error.message);
        }
        return false;
    }
}

export async function getServiceWorkerState() {
    if (!isServiceWorkerSupported()) {
        return null;
    }

    try {
        const registration = await navigator.serviceWorker.getRegistration();
        if (registration) {
            if (registration.active) {
                return registration.active.state;
            } else if (registration.waiting) {
                return 'waiting';
            } else if (registration.installing) {
                return 'installing';
            }
        }
        return 'not-registered';
    } catch (error) {
        console.error('Failed to get service worker state:', error);
        return null;
    }
}

export async function setCachePolicies(policies) {
    // Send cache policies to the active service worker
    if ('serviceWorker' in navigator && navigator.serviceWorker.controller) {
        navigator.serviceWorker.controller.postMessage({
            type: 'SET_CACHE_POLICIES',
            policies: policies
        });
    } else {
        // Wait for the service worker to become active
        navigator.serviceWorker.ready.then(reg => {
            if (reg.active) {
                reg.active.postMessage({
                    type: 'SET_CACHE_POLICIES',
                    policies: policies
                });
            }
        });
    }
}

export async function setOfflineFallbackPath(path) {
    if ('serviceWorker' in navigator && navigator.serviceWorker.controller) {
        navigator.serviceWorker.controller.postMessage({
            type: 'SET_OFFLINE_FALLBACK_PATH',
            path: path
        });
    } else {
        navigator.serviceWorker.ready.then(reg => {
            if (reg.active) {
                reg.active.postMessage({
                    type: 'SET_OFFLINE_FALLBACK_PATH',
                    path: path
                });
            }
        });
    }
}

function setupRegistrationEventListeners(registration) {
    if (registration.installing) {
        setupServiceWorkerEventListeners(registration.installing);
    }
    
    if (registration.waiting) {
        setupServiceWorkerEventListeners(registration.waiting);
    }
    
    if (registration.active) {
        setupServiceWorkerEventListeners(registration.active);
    }
    
    registration.addEventListener('updatefound', () => {
        const newWorker = registration.installing;
        if (newWorker) {
            setupServiceWorkerEventListeners(newWorker);
        }
    });
}

function setupServiceWorkerEventListeners(serviceWorker) {
    serviceWorker.addEventListener('statechange', () => {
        if (dotNetReference) {
            const eventType = serviceWorker.state === 'activated' ? 'activate' : 
                             serviceWorker.state === 'installed' ? 'install' : 
                             serviceWorker.state;
            dotNetReference.invokeMethodAsync('OnServiceWorkerEvent', eventType, serviceWorker.state);
        }
    });
}

export function dispose() {
    dotNetReference = null;
    registrations.clear();
}

File: src/BlazorPwaKit/ServiceWorkerManager.cs
using Microsoft.JSInterop;
using System.Text.Json;

namespace BlazorPwaKit;

/// <summary>
/// Service worker lifecycle events
/// </summary>
public record ServiceWorkerEvent(string Type, string? Message = null);

/// <summary>
/// Service worker registration state
/// </summary>
public enum ServiceWorkerState
{
    NotRegistered,
    Registering,
    Registered,
    Updating,
    Updated,
    Error
}

/// <summary>
/// Service worker registration information
/// </summary>
public record ServiceWorkerRegistration(
    string Scope,
    ServiceWorkerState State,
    string? ScriptUrl = null,
    string? Error = null
);

/// <summary>
/// Manages service worker lifecycle: registration, updates, and unregistration
/// </summary>
public class ServiceWorkerManager : IAsyncDisposable
{
    private readonly IJSRuntime _jsRuntime;
    private readonly Lazy<Task<IJSObjectReference>> _moduleTask;
    private readonly Dictionary<string, ServiceWorkerRegistration> _registrations = new();
    private ICachePolicyProvider? _cachePolicyProvider;

    public event EventHandler<ServiceWorkerEvent>? ServiceWorkerInstalled;
    public event EventHandler<ServiceWorkerEvent>? ServiceWorkerActivated;
    public event EventHandler<ServiceWorkerEvent>? ServiceWorkerFetch;
    public event EventHandler<ServiceWorkerEvent>? ServiceWorkerMessage;
    public event EventHandler<ServiceWorkerEvent>? ServiceWorkerError;
    public event EventHandler<ServiceWorkerEvent>? ServiceWorkerUpdated;

    public ServiceWorkerManager(IJSRuntime jsRuntime)
    {
        _jsRuntime = jsRuntime;
        _moduleTask = new Lazy<Task<IJSObjectReference>>(async () =>
        {
            var module = await _jsRuntime.InvokeAsync<IJSObjectReference>("import", "./_content/BlazorPwaKit/serviceWorkerManager.js");
            await module.InvokeVoidAsync("initialize", DotNetObjectReference.Create(this));
            return module;
        });
    }

    /// <summary>
    /// Registers a service worker with the specified script URL and scope
    /// </summary>
    /// <param name="scriptUrl">The URL of the service worker script</param>
    /// <param name="scope">The scope of the service worker (optional)</param>
    /// <returns>True if registration was successful</returns>
    public async Task<bool> RegisterAsync(string scriptUrl, string? scope = null)
    {
        try
        {
            var key = GetRegistrationKey(scriptUrl, scope);
            _registrations[key] = new ServiceWorkerRegistration(scope ?? "/", ServiceWorkerState.Registering, scriptUrl);
            
            var module = await _moduleTask.Value;
            var result = await module.InvokeAsync<bool>("registerServiceWorker", scriptUrl, scope);
            
            if (result)
            {
                _registrations[key] = _registrations[key] with { State = ServiceWorkerState.Registered };
                return true;
            }
            else
            {
                _registrations[key] = _registrations[key] with { State = ServiceWorkerState.Error, Error = "Registration failed" };
                return false;
            }
        }
        catch (Exception ex)
        {
            var key = GetRegistrationKey(scriptUrl, scope);
            _registrations[key] = new ServiceWorkerRegistration(scope ?? "/", ServiceWorkerState.Error, scriptUrl, ex.Message);
            ServiceWorkerError?.Invoke(this, new ServiceWorkerEvent("error", ex.Message));
            return false;
        }
    }

    /// <summary>
    /// Updates the service worker registration
    /// </summary>
    /// <param name="scriptUrl">The URL of the service worker script</param>
    /// <param name="scope">The scope of the service worker (optional)</param>
    /// <returns>True if update was successful</returns>
    public async Task<bool> UpdateAsync(string scriptUrl, string? scope = null)
    {
        try
        {
            var key = GetRegistrationKey(scriptUrl, scope);
            if (_registrations.TryGetValue(key, out var registration))
            {
                _registrations[key] = registration with { State = ServiceWorkerState.Updating };
            }
            
            var module = await _moduleTask.Value;
            var result = await module.InvokeAsync<bool>("updateServiceWorker", scriptUrl, scope);
            
            if (result && _registrations.TryGetValue(key, out registration))
            {
                _registrations[key] = registration with { State = ServiceWorkerState.Updated };
                ServiceWorkerUpdated?.Invoke(this, new ServiceWorkerEvent("updated"));
                return true;
            }
            
            return false;
        }
        catch (Exception ex)
        {
            ServiceWorkerError?.Invoke(this, new ServiceWorkerEvent("error", ex.Message));
            return false;
        }
    }

    /// <summary>
    /// Unregisters a service worker
    /// </summary>
    /// <param name="scriptUrl">The URL of the service worker script</param>
    /// <param name="scope">The scope of the service worker (optional)</param>
    /// <returns>True if unregistration was successful</returns>
    public async Task<bool> UnregisterAsync(string scriptUrl, string? scope = null)
    {
        try
        {
            var module = await _moduleTask.Value;
            var result = await module.InvokeAsync<bool>("unregisterServiceWorker", scriptUrl, scope);
            
            if (result)
            {
                var key = GetRegistrationKey(scriptUrl, scope);
                _registrations.Remove(key);
                return true;
            }
            
            return false;
        }
        catch (Exception ex)
        {
            ServiceWorkerError?.Invoke(this, new ServiceWorkerEvent("error", ex.Message));
            return false;
        }
    }

    /// <summary>
    /// Gets the current registration information for a service worker
    /// </summary>
    /// <param name="scriptUrl">The URL of the service worker script</param>
    /// <param name="scope">The scope of the service worker (optional)</param>
    /// <returns>Service worker registration information</returns>
    public ServiceWorkerRegistration? GetRegistration(string scriptUrl, string? scope = null)
    {
        var key = GetRegistrationKey(scriptUrl, scope);
        return _registrations.TryGetValue(key, out var registration) ? registration : null;
    }

    /// <summary>
    /// Gets all current service worker registrations
    /// </summary>
    /// <returns>Dictionary of all registrations</returns>
    public IReadOnlyDictionary<string, ServiceWorkerRegistration> GetAllRegistrations()
    {
        return _registrations.AsReadOnly();
    }

    /// <summary>
    /// Checks if service workers are supported in the current browser
    /// </summary>
    /// <returns>True if service workers are supported</returns>
    public async Task<bool> IsServiceWorkerSupportedAsync()
    {
        try
        {
            var module = await _moduleTask.Value;
            return await module.InvokeAsync<bool>("isServiceWorkerSupported");
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Gets the current service worker state
    /// </summary>
    /// <returns>Current service worker state</returns>
    public async Task<string?> GetServiceWorkerStateAsync()
    {
        try
        {
            var module = await _moduleTask.Value;
            return await module.InvokeAsync<string?>("getServiceWorkerState");
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Pushes all cache policies to the service worker
    /// </summary>
    public async Task PushCachePoliciesAsync()
    {
        if (_cachePolicyProvider == null) return;
        var policies = _cachePolicyProvider.ExportPoliciesForJs();
        var module = await _moduleTask.Value;
        await module.InvokeVoidAsync("setCachePolicies", policies);
    }

    /// <summary>
    /// Sets the cache policy provider
    /// </summary>
    /// <param name="provider">The cache policy provider</param>
    public void SetCachePolicyProvider(ICachePolicyProvider provider)
    {
        _cachePolicyProvider = provider;
    }

    /// <summary>
    /// Callback method invoked from JavaScript when service worker events occur
    /// </summary>
    [JSInvokable]
    public void OnServiceWorkerEvent(string eventType, string? message = null)
    {
        var serviceWorkerEvent = new ServiceWorkerEvent(eventType, message);
        
        switch (eventType.ToLowerInvariant())
        {
            case "install":
                ServiceWorkerInstalled?.Invoke(this, serviceWorkerEvent);
                break;
            case "activate":
                ServiceWorkerActivated?.Invoke(this, serviceWorkerEvent);
                break;
            case "fetch":
                ServiceWorkerFetch?.Invoke(this, serviceWorkerEvent);
                break;
            case "message":
                ServiceWorkerMessage?.Invoke(this, serviceWorkerEvent);
                break;
            case "error":
                ServiceWorkerError?.Invoke(this, serviceWorkerEvent);
                break;
            case "updated":
                ServiceWorkerUpdated?.Invoke(this, serviceWorkerEvent);
                break;
        }
    }

    private static string GetRegistrationKey(string scriptUrl, string? scope)
    {
        return $"{scriptUrl}:{scope ?? "/"}";
    }

    public async ValueTask DisposeAsync()
    {
        if (_moduleTask.IsValueCreated)
        {
            var module = await _moduleTask.Value;
            await module.InvokeVoidAsync("dispose");
            await module.DisposeAsync();
        }
    }
}